processing/src/main/java/org/apache/druid/query/TableDataSource.java

```diff
@@ -98,7 +98,7 @@ public final boolean equals(Object o)
     if (this == o) {
       return true;
     }
-    if (!(o instanceof TableDataSource)) {
+    if (!(o instanceof TableDataSource) || !getClass().equals(o.getClass())) {
```

Comment:
This change is causing `SegmentMetadataQueryTest.testSerde()` to fail: https://travis-ci.org/github/apache/druid/jobs/697612764#L5001

The `query` variable has a `LegacyDataSource` whereas the deserialized serialized version has a `TableDataSource`.